vital/types/metadata.cxx

```diff
@@ -165,6 +164,16 @@ ::add( metadata_item* item )
   this->m_metadata_map[item->tag()] = item_ptr(item);
 }
 
+void
+metadata
+::add( std::shared_ptr<metadata_item const> const& item )
```

Comment:
Yike. This interface bugs me... I understand the need, but it would be nice if this was less error-prone. Would it be possible to make the other method take a `unique_ptr` instead of a raw pointer so that the ownership semantics are less subject to accidental misuse? (This would also prevent passing the address of an item allocated on the stack.)

Reply:
That's an interesting approach. Let me see how it works.

Reply:
Working on it :slightly_smiling_face:. Busy falling down the rabbit hole of `vital::any`'s complete lack of move semantics...

Reply:
Okay... this should only need slight tweaking; will need to wrap the cloned value in a `unique_ptr`, but otherwise looks fine.

Although, for simplicity, you might want to consider instead:
```c++
void
metadata
::add( std::unique_ptr< metadata_item >&& item )
{
  if ( !item )
  {
    throw std::invalid_argument{ "null pointer" };
  }

  this->m_metadata_map[ item->tag() ] = item_ptr{ item->clone() };
}
```

(Same effect, but without constructing a temporary `unique_ptr`... which the compiler may elide anyway, the point is more for to simplify things for someone reading the code. The `nullptr` check is optional. I elided it, but probably keep your comment also.)